Dockerfile

```diff
@@ -33,13 +33,25 @@ RUN apt-get update && apt-get -y install software-properties-common \
         curl \
         sudo \
         git \
-    && make deps-ubuntu \
-    && python3 -m venv /usr/local \
+    && make deps-ubuntu
+RUN python3 -m venv /usr/local \
     && hash -r \
     && make install \
-    && eval $FIXUP \
-    && rm -rf /build-ocrd
+    && eval $FIXUP
 
 WORKDIR /data
 
 CMD ["/usr/local/bin/ocrd", "--help"]
+
+FROM ocrd_core_base as ocrd_core_test
+WORKDIR /build-ocrd
+COPY Makefile .
+RUN make assets
+COPY tests ./tests
+COPY .gitmodules .
+COPY requirements_test.txt .
+RUN pip install -r requirements_test.txt
+RUN mkdir /ocrd-data && chmod 777 /ocrd-data
+
+CMD ["yes"]
+# CMD ["make", "test", "integration-test"]
```

Comment:
Can someone explain this to me? My understanding currently is that the core image is changed with this and now pip contains the requirements-test and executes `yes` by default. How can the core-Image still be build with that?

Reply:
My understanding is that the `ocrd_core_base` layer that was built above is cached and continues from there to make assets, install test requirements, and so on. This is useful when only the test code is changed, it's enough to rebuild just the `ocrd_core_test`. However, if the source code changes as well, then `ocrd_core_base` is affected as well and is rebuilt again.

Reply:
But what comes out at the bottom is the core image containing the test stuff. Previously ocrd-core was build with this Dockerfile. But now when it is used to build core the final image contains all the unwanted test code (unwanted in production). I would think we need an extra Dockerfile for the Tests and one for creating the ocrd-core image (the state of the dockerfile before the PR).

Reply:
We're using the multi-stage build feature https://docs.docker.com/build/building/multi-stage/

Essentially:
- Every `FROM` line starts a new stage
- Stage name is set via `FROM ... AS stage-name`
- The target stage can be provided with `--target stage-name`

In the `make docker` makefile target, we specify

```make
docker docker-cuda:                                                                                                                                     
  docker build --progress=plain -f $(DOCKER_FILE) -t $(DOCKER_TAG) --target ocrd_core_base --build-arg BASE_IMAGE=$(DOCKER_BASE_IMAGE) $(DOCKER_ARGS) . 
```

So, it only builds the `ocrd_core_base` stage, which is unchanged. So no need for two Dockerfiles.

Reply:
Thank you